### PR TITLE
Fixes to cdap-defaults.xml

### PR DESCRIPTION
**Pull request by awholegunch**
Addresses https://issues.cask.co/browse/CDAP-4803 (update description for app.artifact.dir)

Passes [Quick Build 6](http://builds.cask.co/browse/CDAP-DOB103-6)

Updates 
- cdap-common/src/main/resources/cdap-default.xml
- cdap-distributions/src/etc/cdap/conf.dist/cdap-site.xml.example
- cdap-examples/Purchase/src/test/resources/cdap-site.xml
- cdap-standalone/src/main/resources/cdap-site.xml

The Documentation Tool (cdap-docs/tools/doc-cdap-default.py) has a new option, "-a", that compares all the files or URLs listed in the file ``cdap-docs/tools/cdap-xml-files.txt`` against the default (cdap-default.xml) and publishes to standard out a report detailing any differences in the descriptions that it finds. It identified a number of properties that are only in the other sources, plus differences in the descriptions.

Fixed formatting issues and two spelling mistakes in the default file.

Added code to the formatter for the documentation to break long lines (if possible) after 40 characters so the table maintains a consistent width down the page. See the revised appendix, where properties and values are sometime split at a period:

Revised Appendices: http://builds.cask.co/artifact/CDAP-DOB103/shared/build-6/Docs-HTML/3.3.1-SNAPSHOT/en/admin-manual/appendices/cdap-site.html

### CODE REVIEW

File: cdap-common/src/main/resources/cdap-default.xml [hunk 1]
```diff
@@ -167,9 +167,10 @@
 
   <property>
     <name>zookeeper.quorum</name>
-    <value>127.0.0.1:2181/${root.namespace}</value>
+    <value>FQDN1:2181,FQDN2:2181/${root.namespace}</value>
```
Comment (dereklwood) [inline cdap-common/src/main/resources/cdap-default.xml]: This changes our default config from one that might work on a dev setup, to one that is always invalid (and possibly java might think it's malformed?).  I don't think we should do this.  These "substitute-me" tokens are only appropriate for example files such as https://github.com/caskdata/cdap/blob/develop/cdap-distributions/src/etc/cdap/conf.dist/cdap-site.xml.example.  Also update description accordingly.

File: cdap-common/src/main/resources/cdap-default.xml [hunk 2]
```diff
@@ -1233,7 +1236,7 @@
     <name>router.bind.address</name>
     <value>0.0.0.0</value>
     <description>
-      Router server address
+      CDAP Router bind address
```
Comment (dereklwood) [inline cdap-common/src/main/resources/cdap-default.xml]: we should be consistent across the main CDAP processes (kafka, router, master, auth, ui).  We have ``Kafka service bind address`` vs ``CDAP Router bind address`` vs ``CDAP Authentication Server bind address``.  Compare ``CDAP`` or not, ``service``/``server`` or not

**Comment by dereklwood** [comment]
couple minor comments on the configs.  I did not review the python code.

File: cdap-common/src/main/resources/cdap-default.xml (hunk 1 above)
Comment (awholegunch) [inline cdap-common/src/main/resources/cdap-default.xml]: Changed to do both: keep the default config to work on a dev setup, but move the change to the description.

File: cdap-docs/tools/cdap-xml-files.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Line starting with a # or space are ignored
+# Support for JSON files to be added
+
+cdap-distributions/src/etc/cdap/conf.dist/cdap-site.xml.example
+cdap-examples/Purchase/src/test/resources/cdap-site.xml
+cdap-standalone/src/main/resources/cdap-site.xml
+# cdap-ui/server/config/development/cdap.json
```
Comment (anwar6953) [inline cdap-docs/tools/cdap-xml-files.txt]: why comment this out?

File: cdap-docs/tools/doc-cdap-default.py [hunk 4]
```diff
@@ -524,15 +587,114 @@ def compare_xml(source, other_source, target):
...
+        item_description = items_dict[name].description
+        other_item_description = other_items_dict[name].description
+        if item_description != other_item_description:
+            print "  Item '%s' does not match" % name
+#             print "    Source:\n    %s" % item_description
+#             print "    Other_source:\n    %s" % other_item_description
+            mis_matches += 1
+        else:
+            matches += 1
+    print "Description matches:     %d" % matches
```
Comment (anwar6953) [inline cdap-docs/tools/doc-cdap-default.py]: wouldn't it be too verbose to print all matches?

File: cdap-docs/tools/doc-cdap-default.py [hunk 5]
```diff
@@ -524,15 +587,114 @@ def compare_xml(source, other_source, target):
...
+    print
+
+def compare_xml_sdl(source, other_source, target=None):
+    # Compares an XML source to an SDL source
+    items = load_xml_source('XML source', source)
+    
+    sdl_source_title = 'SDL other source'
+    print "Loading '%s': %s" % (sdl_source_title, other_source)
+    f, f_source= download_to_file(other_source)
+    import json
```
Comment (anwar6953) [inline cdap-docs/tools/doc-cdap-default.py]: move imports to top of file?

File: cdap-docs/tools/doc-cdap-default.py [hunk 6]
```diff
@@ -101,6 +109,34 @@ def encode(value):
...
+    @staticmethod
+    def _split_text(text, length=70, delimiter='.'):
+        # Splits text into blocks not exceeding length if possible
+        if len(text) > length and text.count(delimiter):
+            # Split text into multiple pieces
+            split = text.rfind('.', 0, length)
+            left = text[:split+1]
+            right = text[split+1:]
+            if right:
+                texts = Item._split_text(right, delimiter)
```
Comment (anwar6953) [inline cdap-docs/tools/doc-cdap-default.py]: This seems like improper usage of the function. It's passing `delimiter` in as the second parameter, but in the function definition, `delimiter` is the third parameter; `length` is the second parameter in the function definition.

Comment (awholegunch) [inline cdap-docs/tools/doc-cdap-default.py]: Good catch. It should also pass the length in the recursive call.

File: cdap-docs/tools/doc-cdap-default.py (hunk 5 above)
Comment (awholegunch) [inline cdap-docs/tools/doc-cdap-default.py]: Sure.

File: cdap-docs/tools/doc-cdap-default.py (hunk 4 above)
Comment (awholegunch) [inline cdap-docs/tools/doc-cdap-default.py]: It's only printing the number of matches.

File: cdap-docs/tools/cdap-xml-files.txt (hunk 3 above)
Comment (awholegunch) [inline cdap-docs/tools/cdap-xml-files.txt]: I would like to add this file to the tests. But I haven't written the support for JSON yet, as noted on line 2.

File: cdap-docs/tools/doc-cdap-default.py (hunk 4 above)
Comment (awholegunch) [inline cdap-docs/tools/doc-cdap-default.py]: But I should remove those two print statements 603 & 604

File: cdap-docs/tools/doc-cdap-default.py [hunk 7]
```diff
@@ -101,6 +110,34 @@ def encode(value):
         else:
             return ''
 
+    @staticmethod
+    def _split_text(text, length=70, delimiter='.'):
+        # Splits text into blocks not exceeding length if possible
+        if len(text) > length and text.count(delimiter):
+            # Split text into multiple pieces
+            split = text.rfind('.', 0, length)
```
Comment (anwar6953) [inline cdap-docs/tools/doc-cdap-default.py]: Seems like the delimiter is not being used, but rather a '.' is being used here.
This is improper, right?

Comment (awholegunch) [inline cdap-docs/tools/doc-cdap-default.py]: Another great catch. Fixed.

File: cdap-docs/tools/cdap-xml-files.txt (hunk 3 above)
Comment (awholegunch) [inline cdap-docs/tools/cdap-xml-files.txt]: Removed.

File: cdap-docs/tools/doc-cdap-default.py [hunk 8]
```diff
@@ -524,15 +588,138 @@ def compare_xml(source, other_source, target):
...
+    other_items.sort(key = lambda p: p.name)
+    other_items_keys = []
+    other_items_dict = {}
+    for item in other_items:
+        other_items_keys.append(item.name)
+        other_items_dict[item.name] = item
+    other_items = (sdl_source_title, other_items, other_items_keys, other_items_dict)
+    compare_items(items, other_items)
+
+def compare_xml_json(source, other_source, target=None):
```
Comment (anwar6953) [inline cdap-docs/tools/doc-cdap-default.py]: to clarify, this method doesn't fully work, yet?

Comment (anwar6953) [inline cdap-docs/tools/doc-cdap-default.py]: If so, maybe we shouldn't utilize it in `compare_all_xml`?

Comment (awholegunch) [inline cdap-docs/tools/doc-cdap-default.py]: ok: let's just remove it.

**Comment by anwar6953** [comment]
LGTM

**Pull request merged**